File: knapsack01.cpp
//Time complexity:  O(Weight*number of items)
//space complexity: O(Weight*number of items)
#include<iostream>
using namespace std;

//long long int max(long long int a, long long int b) {(a > b)? return a : return b; } 
int knapsack(int n,int maxi,int wt[],int value[])
{
long long int k[n+1][maxi+1];
long long int i,w;
for(i=0;i<=n;i++)
{
    for(w=0;w<=maxi;w++)
    {
        if(i==0||w==0)
        {
        k[i][w]=0;}
        else if(wt[i-1]<=w){
        k[i][w]=max(value[i-1]+k[i-1][w-wt[i-1]],k[i-1][w]);}
        else{
        k[i][w]=k[i-1][w];}
    }
}

return k[n][maxi];
}

int main() {
	int test,n,maxi;
	cin>>test;
	for(int i=0;i<test;i++)
	{
	    cin>>n>>maxi;
	    int wt[n],value[n];
	    for(int j=0;j<n;j++)
	    {
	        cin>>value[j];
	    }
	     for(int k=0;k<n;k++)
	    {
	        cin>>wt[k];
	    }
	    cout<<knapsack(n,maxi,wt,value)<<endl;
	}
}

File: knapSack01.java
//Time complexity: O(m*n) | m is the length of the array and n is the capacity(max Weight)
//Space Complexity: O(m*n) | Here I have assumed the maximum length of the array and the maximum weight to be 100 but in normal case, the space complexity should be m*n
/*
    Algorithm: 1. Recursively select each of the possible paths to determine the maximum value
               2. Memoize the recursive calls to optimize the solution
*/
import java.util.Arrays;

public class knapSack01 {
    public static void main(String[] args) {
        int[] wt = {1,3,4,5};
        int[] val = {1,4,5,7};
        setArray();
        System.out.println(knapSack(wt,val,wt.length,7));
    }
    static int[][] dp = new int[100][100];
    private static void setArray(){
        for(int[] temp: dp){
            Arrays.fill(temp,-1);   
        }
    }
    public static int knapSack(int[] wt, int[] val, int n, int W){
        //base case
        if(n == 0 || W == 0) return 0;
        if(dp[n][W] != -1 ) return dp[n][W];
        //Choices
        if(wt[n-1] <= W){
            dp[n][W] = Math.max(val[n - 1] + knapSack(wt, val, n - 1, W - wt[n - 1]), knapSack(wt, val, n - 1, W));
            return dp[n][W];
        }else{
            dp[n][W] = knapSack(wt, val, n - 1, W);
            return  dp[n][W];
        }
    }
}
